### PR TITLE
Add factories for SharedTree lists and maps

### PR DESCRIPTION
**Pull request by noencke**
## Description

This adds the ability to construct lists and maps via `new ...` in the class-tree layer. This does not add `.create` for lists and maps in the simple-tree layer.

## Reviewer Guidance

* Raw nodes are implemented as FlexTreeNodes (including raw object nodes, which have been re-implemented). Much of the implementation is stubbed out since it is not needed, but if/when we decide to allow reading and writing to raw proxies, it can be implemented by filling out those classes/interfaces.
* Proxy content extraction has changed structure a little bit. Now that extraction and hydration is occurring for all three types of nodes rather than just objects, some common behavior has been factored out into helper methods.

### CODE REVIEW

File: experimental/dds/tree2/src/simple-tree/proxies.ts [hunk 1]
```diff
@@ -1011,6 +1067,46 @@ function extractContentObject<T extends object>(input: T): ExtractedFactoryConte
...
+		const factoryContent = extractRawNodeContent(rawEditNode);
+		if (factoryContent === undefined) {
+			// We were passed a proxy, but that proxy doesn't have any raw content.
+			throw new Error("Cannot insert a node that is already in the tree");
+		}
+		content = factoryContent as T;
+		fromFactory = true;
+	}
+
+	// TODO: This checking could be made stronger by using `class-tree/Tree.schema(content).kind`
```
Comment (noencke) [inline experimental/dds/tree2/src/simple-tree/proxies.ts:1092]: FYI @CraigMacomber (it did work when I tried it).

Comment (CraigMacomber) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:1092]: Can you make this work without a cycle if you directly import treeAPI from class-tree rather than its index file?

Comment (CraigMacomber) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:1092]: Since we will be merging the two directories, doing that kind of import to avoid cycles should be fine and is done in a few spots already.

File: experimental/dds/tree2/src/simple-tree/proxies.ts [hunk 2]
```diff
@@ -815,23 +815,66 @@ function createMapProxy<TSchema extends MapNodeSchema>(
...
+	let editNode: RawTreeNode<TSchema, InsertableTypedNode<TreeNodeSchema>>;
+	let proxy: TreeNode;
+	if (schema instanceof ObjectNodeSchema) {
+		const contentCopy = copyContent(schema.name, content as InsertableTypedNode<typeof schema>);
+		editNode = createRawNode(schema, contentCopy);
+		proxy = createObjectProxy(schema, allowAdditionalProperties, target);
+	} else if (schema instanceof FieldNodeSchema) {
+		const contentCopy = copyContent(schema.name, content as InsertableTypedNode<typeof schema>);
+		editNode = createRawNode(schema, contentCopy);
+		proxy = createListProxy(allowAdditionalProperties, target);
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/proxies.ts:855]: Nit: might be worth a quick comment noting why `FieldNodeSchema` implies lists here

File: experimental/dds/tree2/src/simple-tree/proxies.ts [hunk 3]
```diff
@@ -815,23 +815,66 @@ function createMapProxy<TSchema extends MapNodeSchema>(
...
+		proxy = createMapProxy(allowAdditionalProperties, target);
+	} else {
+		fail("Unrecognized content schema");
+	}
+
 	return setEditNode(proxy, editNode);
 }
 
+function copyContent<T extends object>(typeName: TreeNodeSchemaIdentifier, content: T): T {
+	const copy =
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/proxies.ts:868]: Nit and possibly just Josh opinion: an if-elseif-else pattern might be nicer here. Would give a place to document assumptions about the input types and make it a bit more obvious where new type support would need to be slotted in in the future. Curious what others think though. Feel free to disregard :)

File: experimental/dds/tree2/src/simple-tree/proxies.ts [hunk 4]
```diff
@@ -871,37 +914,70 @@ interface ExtractedFactoryContent<T extends InsertableTypedNode<TreeNodeSchema>>
  * }
  * ```
  */
+export function extractFactoryContent<
+	T extends InsertableTypedNode<FieldNodeSchema> | Unhydrated<FieldNodeSchema>,
+>(input: T, insertedAtIndex: number): ExtractedFactoryContent<T>;
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/proxies.ts:919]: Probably worth updating the function comment with an `@param` comment for `insertedAtIndex`

File: experimental/dds/tree2/src/simple-tree/rawNode.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+	TreeNodeSchema,
+	TreeNodeSchemaBase,
+	TreeStatus,
+	FlexTreeFieldNode,
+} from "../feature-libraries";
+import { fail } from "../util";
+import { InsertableTypedNode } from "./insertable";
+
+/** Stores the content of the raw node, i.e. the data that was passed to the factory */
+const nodeContent = Symbol();
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/rawNode.ts:35]: Nit: This variable name feels a bit misleading to me. IMO `nodeContentSymbol` would make it a bit less confusing. But if this better matches the conventions we've been using, please disregard.

File: experimental/dds/tree2/src/simple-tree/proxies.ts [hunk 6]
```diff
@@ -1011,6 +1067,46 @@ function extractContentObject<T extends object>(input: T): ExtractedFactoryConte
...
+	if (isFluidHandle(content)) {
+		type = "leaf";
+	} else if (isReadonlyArray(content)) {
+		type = "list";
+	} else if (content instanceof Map) {
+		type = "map";
+	} else if (content !== null && typeof content === "object") {
+		type = "object";
+	} else {
+		type = "leaf";
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/proxies.ts:1104]: Nit: should we assert anything here?

File: experimental/dds/tree2/src/simple-tree/rawNode.ts [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+interface HasNodeContent<T> {
+	[nodeContent]: T;
+}
+
+/**
+ * Retrieve the content of a raw object node created via {@link createRawNode}.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/rawNode.ts:41]: Nit: probably worth noting that this actually removes the content from the input object

File: experimental/dds/tree2/src/simple-tree/rawNode.ts [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+		const content = node[nodeContent] ?? fail("Node content may only be extracted once");
+		Reflect.deleteProperty(node, nodeContent);
+		return content as object;
+	}
+
+	return undefined;
+}
+
+/**
+ * Creates a node that falsely pretends to satisfy the given schema while wrapping the given node content.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/simple-tree/rawNode.ts:69]: Nit (I think "pretends" is sufficient to convey the falsness): 
```suggestion
 * Creates a node that pretends to satisfy the given schema while wrapping the given node content.
```

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 1 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:1098]: It did not like the import to `../class-tree/treeAPi`, but it is okay with imports from `"../class-tree/schemaTypes"`, so I got it working by adding a helper at the bottom of this file (see `getNodeKind`).

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 3 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:871]: I am fine with if-else, but then I would either have to make `copy` be a `let`, because it is used below all cases, or copy the Object.define property into every case. Not the end of the world, but neither are double-ternaries IMO (so long as prettier does a good job, like this), so I'll just leave it for now. If the type checking gets more complicated in the future then I could see it being obviously better to do if/else.

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.spec.ts [hunk 9]
```diff
@@ -262,7 +262,7 @@ describe("schemaFactory", () => {
 		const view: TreeView<Canvas> = tree.schematize(config);
 		const stuff = view.root.stuff;
 		assert(stuff instanceof NodeList);
-		const item = stuff[1];
+		const item = stuff[0];
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/class-tree/schemaFactory.spec.ts:265]: Unrelated bug fix?

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 2 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:858]: Okay

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 4 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:924]: Sure!

File: experimental/dds/tree2/src/simple-tree/proxies.ts [hunk 10]
```diff
@@ -1011,6 +1067,46 @@ function extractContentObject<T extends object>(input: T): ExtractedFactoryConte
...
+		content = factoryContent as T;
+		fromFactory = true;
+	}
+
+	// TODO: This checking could be made stronger by using `class-tree/Tree.schema(content).kind`
+	// to get the type when `content` is a proxy. At the moment, importing `Tree.schema` causes a cycle.
+	let type: "object" | "list" | "map" | "leaf";
+	if (isFluidHandle(content)) {
+		type = "leaf";
+	} else if (isReadonlyArray(content)) {
```
[the remark this answers is not in the document]
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts]: Is there something I ought to do to investigate this further?

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 6 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:1110]: What do you have in mind?

File: experimental/dds/tree2/src/simple-tree/rawNode.ts (hunk 5 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/rawNode.ts:35]: I don't like the convention of putting `Symbol` on the end of symbols, it makes them too long and ugly. I don't think we have an established pattern - this is replacing a symbol that was also named `nodeContent`.

File: experimental/dds/tree2/src/simple-tree/rawNode.ts (hunk 8 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/rawNode.ts]: Good call.

Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/rawNode.ts]: I supposed "falsely pretending" might actually imply that it's doing the real thing, correctly lol

File: experimental/dds/tree2/src/simple-tree/rawNode.ts [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+ * Retrieve the node content via {@link extractRawNodeContent}.
+ *
+ * @remarks This is useful for creating "raw" nodes: nodes which capture data about a pending insertion but are not yet inserted.
```
[the remark this answers is not in the document]
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/rawNode.ts:72]: Woo! When Josh says docs are great, you _know_ they're great :D

File: experimental/dds/tree2/src/simple-tree/rawNode.ts (hunk 7 above)
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/rawNode.ts:41]: I clarified the remarks.

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.spec.ts (hunk 9 above)
Comment (noencke) [reply experimental/dds/tree2/src/test/class-tree/schemaFactory.spec.ts:265]: Yea - you can see the list being initialized up above with only one thing in it. The test itself is not unrelated though - it was skipped because it needed the functionality that is added in this PR.

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 6 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/simple-tree/proxies.ts:1110]: Sorry, I was originally thinking there might be value in asserting the content was of the shape we expect, since this is the catch-all else case. But after reading the code again, this _has_ to be a primitive or undefined. So disregard 😋

File: experimental/dds/tree2/src/simple-tree/proxies.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (noencke) [reply experimental/dds/tree2/src/simple-tree/proxies.ts]: Whether or not it was working before, it should be better now since we're checking the class node kind, if there is one.

**Review by Josmithr — APPROVED** [review]
